src/dataflow/src/render/mod.rs

```diff
@@ -205,28 +209,47 @@ pub struct RelevantTokens {
     pub cdc_tokens: HashMap<GlobalId, Rc<dyn Any>>,
 }
 
-/// Build a dataflow from a description.
-pub fn build_dataflow<A: Allocate>(
+/// Information about each source that must be communicated between storage and compute layers.
+pub struct SourceBoundary {
+    /// Captured `row` updates representing a differential collection.
+    pub ok: Rc<EventLink<repr::Timestamp, (Row, repr::Timestamp, repr::Diff)>>,
+    /// Captured error updates representing a differential collection.
+    pub err: Rc<EventLink<repr::Timestamp, (DataflowError, repr::Timestamp, repr::Diff)>>,
+    /// A token that should be dropped to terminate the source.
+    pub token: Rc<Option<crate::source::SourceToken>>,
+    /// Additional tokens that should be dropped to terminate the suorce.
+    pub additional_tokens: Vec<Rc<dyn std::any::Any>>,
```

Comment:
I once understood why there are two types of tokens, but, oy, I've forgotten the details. Note to self, I guess, to figure out why `SourceToken` can't get stored in this box of things.

Reply:
Hmm, this diff seems to pass just fine:

```diff
diff --git a/src/dataflow/src/source/mod.rs b/src/dataflow/src/source/mod.rs
index 155ba36ab..9fef9dfe6 100644
--- a/src/dataflow/src/source/mod.rs
+++ b/src/dataflow/src/source/mod.rs
@@ -268,13 +268,6 @@ pub struct SourceToken {
     activator: Activator,
 }
 
-impl SourceToken {
-    /// Re-activates the associated timely source operator.
-    pub fn activate(&self) {
-        self.activator.activate();
-    }
-}
-
 impl Drop for SourceToken {
     fn drop(&mut self) {
         *self.capabilities.borrow_mut() = None;

```

So I extra don't understand why this can't go in the `dyn Any` box.

Reply:
Seriously, though, just musing. No need to take any action here!

Reply:
It may be a historical vestige. I think `additional_tokens` are e.g. CDC tokens, and were put in a different pile to distinguish them.  I'm not sure if we need to know anything about the sources other than how to turn them off.

Reply:
You might think CDC tokens are in `cdc_tokens`, but that appears to be incorrect and that field of `RelevantTokens` is never referenced and can probably be deleted also.